### PR TITLE
[HTTPCLient] Add AccessTokenCreator

### LINKED ISSUE #35: Deal with access token
Many API are accessible only with an access token, we can improve `HTTP client wrappers` to add a way to retrieve (create) an access token and pass it in the request.

We can have a default AccessTokenCreator helper that can be use by default and configured via the config and we can let the possibility to register a CustomAccessTokenCreator with an interface

```php
interface AccessTokenCreatorInterface
getAccessToken()

class DefaultAcessTokenCreator implements AccessTokenCreatorInterface
// configured form config file

class CustomAccessTokenCreator implements AccessTokenCreatorInterface
// Logic implemented by the end user

AbstractMicroserviceHttpRepository::request() 
{
  if($this->accessTokenCreator !== null)
   {
      $this->accessTokenCreator->getAccessToken()
      // pass it into header
}
```
[Edit] :warning: 
Since we deal with tag and autoconfiguration, The AccessTokenCreator helpers have to implement `AccessTokenCreatorInterface` and they can also deal with the priority :tada:

### PR DESCRIPTION
**Pull request by ismail1432**
Should resolve #35 

AccessTokenCreator are autoconfigured and the end user can deal with priority in the service definition, the first  AccessTokenCreator that returns an access token is defined to retrieve an access token and the header

Not sure about : 

~~- the namespace~~
~~- credential in PHP doc~~

- [x] Tests
- [x] Documentation

### CODE REVIEW

**Comment by ismail1432** [comment]
@mtarld I rebase with master, There is not test yet but I'm working on,  BTW I have some question 

- I'm not sure about the namespace it was `Security` and I replace by `AccessToken` and I'm still not sure
- I remove [all service declaration ](https://github.com/mtarld/api-platform-ms-bundle/pull/37/commits/2f4bc98a477bced64a7e2cca8a6cce637af495c6#diff-4a3af44c7eda2b193f97cbbaeb52568de5ce4f6843acb8416a4396ce7981190fL31) it seems useless but I don't know if I keep this change in this PR, WDYT ?

**Comment by mtarld** [comment]
> @mtarld I rebase with master, There is not test yet but I'm working on, BTW I have some question
> 
> * I'm not sure about the namespace it was `Security` and I replace by `AccessToken` and I'm still not sure

Actullay, I'm wondering if the "AccessToken" stuff should simply go in the `HttpClient` namespace as it's just a part of the HTTP Client. WDYT?

File: src/DependencyInjection/ApiPlatformMsExtension.php [hunk 1]
```diff
@@ -63,5 +65,8 @@ public function load(array $configs, ContainerBuilder $container): void
         $container->setParameter('api_platform_ms.hosts', $config['hosts']);
         $container->setParameter('api_platform_ms.microservices', $config['microservices']);
         $container->setParameter('api_platform_ms.enabled_formats', $formats);
+
+        $container->registerForAutoconfiguration(AuthenticationHeaderProviderInterface::class)->addTag('api_platform_ms.authentication_header_provider')
```
Comment (mtarld) [inline src/DependencyInjection/ApiPlatformMsExtension.php:69]:
```suggestion
        $container->registerForAutoconfiguration(AuthenticationHeaderProviderInterface::class)->addTag('api_platform_ms.authentication_header_provider');
```

File: src/Resources/doc/tools/authentication-header-provider.md [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+Even if authentication headers are part of the request headers,
+we don't want to specify them each time we call the `request` method.
+Therefore, we can use authentication header providers to inject authentication headers
+in requests and decouple them from the regular headers.
+
+## Description
+Create your Provider and implements `AuthenticationHeaderProviderInterface`.
+Behind the scene the Http Client will iterate over them and call `AuthenticationHeaderProviderInterface::supports` in order to fetch the good provider.
+
+TODO Talk about tagged iterators, priority and link to the Symfony doc
```
Comment (mtarld) [inline src/Resources/doc/tools/authentication-header-provider.md:16]: I'll give you the content ASAP

File: tests/Fixtures/App/src/Authentication/BasicAuthenticationHeaderProvider.php [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+
+use Mtarld\ApiPlatformMsBundle\HttpClient\AuthenticationHeaderProviderInterface;
+use Mtarld\ApiPlatformMsBundle\Microservice\Microservice;
+
+/**
+ * @author Smaïne Milianni <smaine.milianni@gmail.com>
+ */
+class BasicAuthenticationHeaderProvider implements AuthenticationHeaderProviderInterface
+{
+    private $enable = false;
```
Comment (mtarld) [inline tests/Fixtures/App/src/Authentication/BasicAuthenticationHeaderProvider.php:13]: Could you rename `enable` to `enabled` and `setEnable` to `setEnabled`? (Same goes for `BearerAuthenticationHeaderProvider`)

Comment (mtarld) [inline tests/Fixtures/App/src/Authentication/BasicAuthenticationHeaderProvider.php:13]: Can you add phpdoc as well?

File: tests/HttpClient/HttpClientTest.php [hunk 4]
```diff
@@ -132,4 +134,38 @@ public function testSwitchHttpClient(): void
         $microserviceHttpClient->setWrappedHttpClient($secondHttpClient);
         $microserviceHttpClient->request('GET', '/puppies');
     }
+
+    /**
+     * @dataProvider authenticationHeadersDataProvider
+     */
+    public function testAuthenticationHeaders($microservice, $method, $uri, array $expectedAuthenticationHeader): void
```
Comment (mtarld) [inline tests/HttpClient/HttpClientTest.php:141]:
```suggestion
    public function testAuthenticationHeaders(string $microservice, string $method, string $uri, array $expectedAuthenticationHeader): void
```

**Review by mtarld — CHANGES_REQUESTED** [review]
I really like that PR, the feature will be truly powerful! Many thanks :slightly_smiling_face: 

That's almost good. Some little comments though.

**Comment by ismail1432** [comment]
TY thanks for review I'll do them asap

File: src/Resources/doc/tools/authentication-header-provider.md (hunk 2 above)
Comment (mtarld) [reply src/Resources/doc/tools/authentication-header-provider.md]: Thanks for the doc part, it's perfect :+1:

**Review by mtarld — APPROVED** [review]
LGTM! Many thanks @ismail1432

**Pull request merged**